borsh-rs/borsh/src/io.rs

```diff
@@ -0,0 +1,85 @@
+use std::io::{Error, ErrorKind, Result};
+
+pub trait Input {
+    fn rem_len(&mut self) -> Result<usize>;
+    fn read_byte(&mut self) -> Result<u8>;
+    fn read(&mut self, buf: &mut [u8]) -> Result<()>;
+}
+
+impl Input for &[u8] {
```

Comment:
This interface will not allow us to use byteorder which we want to use since it reduces the number of allocations, as explained in here: https://github.com/nearprotocol/borsh/issues/29#issuecomment-548530050

https://docs.rs/byteorder/1.1.0/byteorder/trait.ByteOrder.html#tymethod.read_u64 takes slice only. I am not sure what is the benefit of depending on `Input` trait instead of depending on the slice, directly.

Reply:
> I am not sure what is the benefit of depending on Input trait instead of depending on the slice, directly.

We need to implement the "read" logic witch would shift the pointer to unread data somewhere. We can call a function on slice like `fn read(mut src: &mut [u8], buf: &mut [u8])`, but I make a trait on type is just more Rust-idiomatic here (there is no other differences).

Reply:
I tried to use just slices and ended up with the need of a second indirection here and also the is a duplication of the logic: https://github.com/nearprotocol/borsh/blob/fckt/read-to-slice2/borsh-rs/borsh/src/de/mod.rs#L10

Reply:
I agree encapsulating a slice into a trait is more idiomatic.